checker.cpp

```diff
@@ -1,17 +1,105 @@
-#include <assert.h>
-
-bool vitalsAreOk(float bpm, float spo2, float respRate) {
-  if(bpm < 70 || bpm > 150) {
-    return false;
-  } else if(spo2 < 90) {
-    return false;
-  } else if(respRate < 30 || respRate > 95) {
-    return false;
-  }
-  return true;
+#include<iostream>
+#include<vector>
+#include<assert.h>
+
+using namespace std;
+
+class AlertInterface
+{
+public:
+	virtual void raiseAlert(const char*, const char*) = 0;
+};
+
+class AlertWithSMS : public AlertInterface
+{
+public:
+	void raiseAlert(const char* vitalName, const char* vitalLevel)
+	{
+		cout << "SMS : " << vitalName << " " << vitalLevel << endl;
+	}
+};
+
+class AlertWithSound : public AlertInterface
+{
+public:
+	void raiseAlert(const char* vitalName, const char* vitalLevel)
+	{
+		cout << "Sound : " << vitalName << " " << vitalLevel << endl;
+	}
+};
+
+class AlertIntegrator : public AlertInterface
+{
+private:
+	AlertWithSMS alertWithSMS;
+	AlertWithSound alertWithSound;
+public:
+	void raiseAlert(const char* vitalName, const char* vitalLevel) override
+	{
+		alertWithSMS.raiseAlert(vitalName, vitalLevel);
+		alertWithSound.raiseAlert(vitalName, vitalLevel);
+	}
+};
+
+class VitalRangeChecker
+{
+private:
+	int lowerLimit;
+	int upperLimit;
+	const char* vitalName;
+	AlertInterface* alerter;
+
+public:
+	VitalRangeChecker(const char*, int, int, AlertInterface*);
+	void checkVitalLevel(float);
+};
+
+VitalRangeChecker::VitalRangeChecker(const char* vitalName, int lowerLimit, int upperLimit, AlertInterface* alerter)
+{
+	this->vitalName = vitalName;
+	this->lowerLimit = lowerLimit;
+	this->upperLimit = upperLimit;
+	this->alerter = alerter;
+}
+
+void VitalRangeChecker::checkVitalLevel(float vitalValue)
+{
+	if (vitalValue < lowerLimit)
+		alerter->raiseAlert(vitalName, "TOO LOW");
+	else if (vitalValue > upperLimit)
+		alerter->raiseAlert(vitalName, "TOO HIGH");
+	else
+		alerter->raiseAlert(vitalName, "OK");
+}
+
+class VitalsIntegrator
+{
+private:
+	vector<VitalRangeChecker> vitalDetails;
+
+public:
+	VitalsIntegrator(AlertInterface* alertPtr)
+	{
+		vitalDetails.push_back(VitalRangeChecker("bpm", 70, 150, alertPtr));
+		vitalDetails.push_back(VitalRangeChecker("spo2", 90, 100, alertPtr));
+		vitalDetails.push_back(VitalRangeChecker("respRate", 30, 95, alertPtr));
+	}
+	void checkAllVitals(const vector<float>);
+};
+
+void VitalsIntegrator::checkAllVitals(vector<float> vitalValuesToCheck)
+{
```

Comment:
Good use of vector without assuming the size. However, there is an assumption on the sequence of vitals. How could we express that a particular vital hasn't been measured?
E.g., in future if we extend this to BP and Sugar, we will add two parameters to the vector, but if a patient is not diabetic and doesn't need Sugar to be monitored, how to express that in the data-structure?
See if you would like to use a map.

Reply:
Hello @sudeeprp , I have replaced the vector with a map. The problem of adding and removal of vitals is also taken care of. And, we can test for any combination of vitals in any order, based on the need.